source/slang/slang-check-overload.cpp

```diff
@@ -358,11 +358,54 @@ namespace Slang
         return true;
     }
 
+    bool isEffectivelyMutating(CallableDecl* decl)
+    {
+        if(decl->hasModifier<MutatingAttribute>())
+            return true;
+
+        if(decl->hasModifier<NonmutatingAttribute>())
+            return false;
+
+        if(as<SetterDecl>(decl))
+            return true;
+
+        return false;
+    }
+
     bool SemanticsVisitor::TryCheckOverloadCandidateDirections(
-        OverloadResolveContext&		/*context*/,
-        OverloadCandidate const&	/*candidate*/)
+        OverloadResolveContext&     context,
+        OverloadCandidate const&    candidate)
     {
-        // TODO(tfoley): check `in` and `out` markers, as needed.
+        if(candidate.flavor != OverloadCandidate::Flavor::Func)
+            return true;
+
+        auto funcDeclRef = candidate.item.declRef.as<CallableDecl>();
+        SLANG_ASSERT(funcDeclRef);
+
+        // Note: This operation was originally introduced as
+        // a place to add checking around l-value-ness of arguments
+        // and parameters, but currently that checking is being
+        // done in other places.
+        //
+        // For now we will only use this step to check the
+        // mutability of the `this` parameter where necessary.
+        //
+        if(!isEffectivelyStatic(funcDeclRef.getDecl()))
+        {
+            if(isEffectivelyMutating(funcDeclRef.getDecl()))
+            {
+                if(context.baseExpr && !context.baseExpr->type.isLeftValue)
+                {
+                    if(context.mode == OverloadResolveContext::Mode::ForReal)
```

Comment:
I wonder if it is possible to also provide additional diagnostic for unsatisfied interface requirement to point out specifically the reason, rather than just return `false` here and print out a general message.

Reply:
Yes, we could use an approach like what we do for overload resolution where we keep track of one or more candidates that are “best” but not necessarily usable. When we fail to find a usable candidate we can take whatever the best candidates are and explain why they weren’t applicable.